### PR TITLE
feat: NvDecoder (H264) for DirectX11

### PR DESCRIPTION
**Pull request by kannan-xiao4**
- Support H264 Decoder using NvDecoder
- Add Encode/Decode native test

Todo
- Check working on Vulkan
- future: Support other input format

### CODE REVIEW

File: Plugin~/WebRTCPlugin/Codec/NvCodec/NvDecoderImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
+#include "pch.h"
+
+#include "../Utils/NvCodecUtils.h"
```
Comment (karasusan) [inline Plugin~/WebRTCPlugin/Codec/NvCodec/NvDecoderImpl.cpp:3]:
```suggestion
#include "NvCodecUtils.h"
```

File: Plugin~/WebRTCPlugin/Codec/NvCodec/NvDecoderImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+#include "NvDecoder/NvDecoder.h"
+#include "NvDecoderImpl.h"
+#include "api/video/i420_buffer.h"
+#include "api/video/video_codec_type.h"
+#include "third_party/libyuv/include/libyuv/convert.h"
+
+using namespace webrtc;
```
Comment (karasusan) [inline Plugin~/WebRTCPlugin/Codec/NvCodec/NvDecoderImpl.cpp:10]: This line is unnecessary.

File: Plugin~/WebRTCPlugin/Codec/NvCodec/NvDecoderImpl.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        NvDecoderImpl(CUcontext context);
+        NvDecoderImpl(const NvDecoderImpl&) = delete;
+        NvDecoderImpl& operator=(const NvDecoderImpl&) = delete;
+        ~NvDecoderImpl() override;
+
+        virtual int32_t InitDecode(const VideoCodec* codec_settings, int32_t number_of_cores) override;
+        virtual int32_t Decode(const EncodedImage& input_image, bool missing_frames, int64_t render_time_ms) override;
+        virtual int32_t RegisterDecodeCompleteCallback(DecodedImageCallback* callback) override;
+        virtual int32_t Release() override;
+        virtual DecoderInfo GetDecoderInfo() const;
```
Comment (karasusan) [inline Plugin~/WebRTCPlugin/Codec/NvCodec/NvDecoderImpl.h:29]:
```suggestion
        virtual DecoderInfo GetDecoderInfo() const override;
```

**Review by karasusan — APPROVED** [review]

**Comment by EricUUU** [comment]
Can I use nvdecoder to decode H264 on unity's windows @karasusan

**Comment by avicarpio** [comment]
This works on linux with version 3.0.0-pre.3 ?

**Pull request merged**